lib/semantics/check-coarray.cc

```diff
@@ -17,6 +17,51 @@
 
 namespace Fortran::semantics {
 
+class CriticalBodyEnforce {
```

Comment:
Consider subclassing semantics::BaseChecker here and using Enter and Leave member functions rather than Pre and Post; the results would require less boilerplate and might be more readable.

Reply:
My understanding is probably poor here.
This class and its usage follow the DoconcurrentBodyEnforce class and its usage in check-do.cc. The parser::Walk function is used to walk the Body of the constructs (Doconcurrent/critical). The parser::Walk function requires Pre and Post function definitions.
The ones with Enter and Leave seem to be called by the SemanticsVisitor class. Are you suggesting to use the SemanticsVisitor class here?
I was initially thinking of creating a new base class, for these kinds of classes, which houses the base Pre and Post functions. This new class can be a base for similar classes like SymbolDumpVisitor in unparse-with-symbols.cc,
DoConcurrentBodyEnforce in check-do.cc etc.

Reply:
That's fine.

Reply:
If you meant fine for the new base class which houses the boiler plate pre and post functions, I can do that in a separate PR.

Reply:
No, I mean just leave it as is.